src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/types/Type.java

```diff
@@ -0,0 +1,89 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types;
+
+import com.amazon.opendistroforelasticsearch.sql.antlr.semantic.SemanticAnalysisException;
+import com.amazon.opendistroforelasticsearch.sql.antlr.visitor.Reducible;
+import com.amazon.opendistroforelasticsearch.sql.utils.StringUtils;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types.BaseType.TYPE_ERROR;
+
+/**
+ * Type interface which represents any type of symbol in the SQL.
+ */
+public interface Type extends Reducible {
+
+    /**
+     * Hide generic type ugliness and error check here in one place.
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    default <T extends Reducible> T reduce(List<T> others) {
```

Comment:
I think the reduce operation only applied to TypeExpression. Is it possible to limit it to TypeExpression only?

Reply:
Essentially `Reducible` and `reduce()` provide a generic interface to visitor so we can implement different logic. For example, we could have a AST generate visitor with AST implement `Reducible`. And we could have another implementation for test case or documentation generating. In one word, anything reducible can implement it here rather than limited to type expression.